### ISSUE TITLE
Distance measures for quantum states

### ISSUE THREAD

**Issue by mpharrigan** [quantumlib/Cirq#2763]
- fidelity
- process fidelity
- purity
- diamond norm
- others?

**Comment by viathor** [comment]
Well, trace distance, of course.

Regarding other distance measures to consider, it is surprisingly easy for them to proliferate. Given a distance measure on quantum states you can define a distance measure on quantum operations by:
1. Applying the former to the (trace-normalized) process matrices.
2. Computing the average over input states.
3. Computing the maximum over input states.

The latter has some cool properties, but is not invariant under tensoring with the identity. The problem can be remedied by a small tweak to the definition (first tensor with the identity on a sufficiently large Hilbert space, then compute maximum).

Now, the diamond norm is the distance measure for quantum processes obtained by applying maximization over input states (with the above tweak) to the trace distance. You can apply the same procedure to other distance measures. At least one of them is actually interesting and that is when you apply maximization over input states (again, with the above tweak) to the fidelity. I don't think this has an established name. I certainly haven't seen anyone refer to this as "diamond fidelity". I have seen it called "stabilized fidelity" or "S fidelity".

Also, since you've included purity, which is not a distance measure, we may consider normalized linear entropy, tangle, Schmidt rank, von Neumann entropy and maybe other measures of entanglement.

So, there is potentially a lot of stuff to do here. However, there are reasons for parsimony. If anything the field suffers from too many rather than too few distance measures and people came up with well-justified recommendations to settle on. We should probably endorse one (or few) of those and stick to them. Moreover, some of distance measures one can find in the literature (e.g. the Bures metric) are simple functions of the fidelity anyway so users who really need them can very easily compute them from the fidelity (trace distance is an exception). 

Also, regarding distance measures for operations, not every combination of procedures 1-3 above and a distance measure for quantum states yields a reasonable distance measure for quantum operations. Process fidelity and diamond norm SGTM. Process trace distance and the S fidelity may be reasonable additions. And we should probably stop there.

Regarding measures of entanglement, many of them are NP-hard to compute.

**Comment by viathor** [comment]
There is another, related concern that came up in a discussion I had with @maffoo. We probably want to avoid developing and maintaining a full quantum information package inside cirq. Many such packages already exist, e.g. [qiskit](https://qiskit.org/) and [qutip](http://qutip.org/) and have a lot more functionality than we are likely to build. Rather than writing our own, we may want to consider partnering with an existing project and ensuring easy interoperability.

@dabacon @Strilanc

**Comment by mpharrigan** [comment]
Love the idea of interop. qiskit and qutip are good, but would each be a very heavy dependency. If we can figure out a way to highlight the quantum information parts, that would be very cool

**Comment by viathor** [comment]
I had quickly skimmed the two packages. They both provide a ton of cool functionality. Here are two short snippets showing the computation of fidelity of two pure states:

qutip:

```python
import qutip

q_0 = qutip.basis(2, 0)
q_1 = qutip.basis(2, 1)
q_plus = (q_0 + q_1).unit()

f = qutip.metrics.fidelity(q_1, q_plus)
print(f)
```

output: 0.7071067811865475

qiskit:

```python
import numpy as np
import qiskit.quantum_info as qi

q_0 = np.array([1, 0])
q_1 = np.array([0, 1])
q_plus = np.array([1, 1]) / np.sqrt(2)

f = qi.state_fidelity(q_0, q_plus)
print(f)
```

output: 0.4999999999999999

Two notable differences:
 * qutip requires that we instantiate their classes that represent quantum states whereas qiskit is happy to work with numpy arrays of amplitudes (which is what cirq produces),
 * qutip uses the older Mike&Ike convention defining fidelity as the square root of <phi|rho|phi>[1] whereas qiskit uses the more modern[2] convention defining fidelity as <phi|rho|phi>.

---
[1]: Assuming one state is pure.
[2]: See paragraph beginning with "Nevertheless..." on p.4 in https://arxiv.org/pdf/quant-ph/0408063.pdf for one recommendation against Mike&Ike's convention.

**Comment by kevinsung** [comment]
See https://github.com/quantumlib/Cirq/pull/2797#issuecomment-593767332 for an example of how Qiskit's implementation of fidelity is much slower than an alternative. If we implement a few core functions ourselves, we can be sure that they are as efficient as possible. Just a few core functions would not take much work to maintain, either.

**Comment by vtomole** [comment]
Rolling our own ~10 procedures from https://qiskit.org/documentation/apidoc/quantum_info.html shouldn't be a huge burden when it comes to maintenance. I vote for wheel re-inventing in this case.

**Comment by tonybruguier-google** [comment]
Hello,
Is more work needed for this issue?
Is this a good reference for the diamond norm?
https://arxiv.org/pdf/1004.4110.pdf

@mpharrigan 
Is this easy enough given my limited experience?

**Comment by mpharrigan** [comment]
I'm not an expert on distance measures or the current state of `cirq.qis`, so someone else could probably speak more to the current state of things. 

But after a quick check, it looks like we still only have (state) fidelity. We should definitely support process fidelity and purity (see @vtomole 's linked qiskit page for more details). This depends on having a representation for processes (like the superoperator representation) 

You're certainly welcome to try. I've found a lot of the most insightful reference material is buried in people's theses like https://uwspace.uwaterloo.ca/handle/10012/9557 which probably contains parts of https://arxiv.org/abs/1111.6950

**Comment by tonybruguier-google** [comment]
Sent https://github.com/quantumlib/Cirq/pull/3198 for process fidelity. Happy to withdraw if this is misguided.

**Comment by tonybruguier** [comment]
@balopat balopat You can assign this issue to me, if you want.

**Comment by balopat** [comment]
@tonybruguier assigned! 
Please outline a design (here should be sufficient) instead of directly opening the PRs directly. I think there are a couple of design questions that are open.

**Comment by tonybruguier** [comment]
Thanks,

There is some discussion going on about how to handle noisy channels and which representation to choose. I plan to wait until there is resolution (I don't think I'm competent to chime in) and then once that decision is made, I will implement some of the functions as per the papers:
https://arxiv.org/pdf/quant-ph/0205035.pdf
https://arxiv.org/pdf/1102.0948.pdf

I already have PR #3226 out for review, but I think it's better for me to wait for now.

**Comment by 95-martin-orion** [comment]
Closed as part of #3641. Please open a new issue if you have a specific measure you're interested in.

**Comment by mpharrigan** [comment]
I'm really confused. You closed this referencing another issue which I'm very unsure why it was closed

**Comment by 95-martin-orion** [comment]
This was discussed in the Cirq sync (and I'm only now seeing the comment on the other issue about transcribing discussions...sorry about that!) but I think the intent was not to mark this as complete, but to remove it from the open items leading into 1.0. Now that we have the `time/after-1.0` tag, we can annotate with that and keep this open.

**Comment by 95-martin-orion** [comment]
Note that `time/after-1.0` was added with the understanding that the general task (adding all distance measures) is not prioritized prior to the 1.0 release, but individual distance measures can still be prioritized by opening focused issues with the `time/before-1.0` tag.

**Comment by mpharrigan** [comment]
got it, thanks!

**Comment by ACE07-Sev** [comment]
Anything left from this issue that needs adding?

**Comment by pavoljuhas** [comment]
Yes, please see the [cirq.qis](https://github.com/quantumlib/Cirq/tree/main/cirq-core/cirq/qis) module to find which ones are already done.
